### PR TITLE
bpo-44490: Improve typing module compatibility with types.Union

### PR DESCRIPTION
**Pull request by uriyyo**
https://bugs.python.org/issue44490

### CODE REVIEW

**Comment by uriyyo** [comment]
@Fidget-Spinner Could you please review this PR?

**Comment by Fidget-Spinner** [comment]
> @Fidget-Spinner Could you please review this PR?

The Python typing parts LGTM. But I need to do some revision for the C parts.

**Comment by Fidget-Spinner** [comment]
Hmm  I wonder if it would be better if you split out the pure Python changes (like the nested `__parameters__` search in typing) from the C+Python changes into a separate PR.

My reasoning follows that the Python changes can easily be backported. The C changes not so much and will require much more deliberation. What do you think Guido? @gvanrossum

**Comment by gvanrossum** [comment]
Agreed, this currently looks like two independent changes. See also the NEWS file.

**Comment by uriyyo** [comment]
@Fidget-Spinner @gvanrossum Thats makes sense. I will convert this PR into containing only `typing` fixes and will open another PR.

**Comment by uriyyo** [comment]
@Fidget-Spinner Could you please review this PR?

File: Lib/typing.py [hunk 1]
```diff
@@ -1775,6 +1777,12 @@ def _strip_annotations(t):
         if stripped_args == t.__args__:
             return t
         return GenericAlias(t.__origin__, stripped_args)
+    if isinstance(t, types.Union):
```
Comment (Fidget-Spinner) [inline Lib/typing.py:1780]: Just double checking, this test case (and the other one in `_eval_type`) is covered by the `get_type_hints` test cases above right?

File: Misc/NEWS.d/next/Core and Builtins/2021-07-06-22-22-15.bpo-44490.BJxPbZ.rst [hunk 2]
```diff
@@ -0,0 +1,2 @@
+Fix ``typing`` module issues with ``types.Union`` parameters and arguments
+resolving. Patch provided by Yurii Karabas.
```
Comment (Fidget-Spinner) [inline Misc/NEWS.d/next/Core and Builtins/2021-07-06-22-22-15.bpo-44490.BJxPbZ.rst]: Seems a little vague, here's a suggestion:
```suggestion
:mod:`typing` now searches for type parameters in ``types.Union`` objects.
``get_type_hints`` will also properly resolve annotations with nested
``types.Union`` objects. Patch provided by Yurii Karabas.
```
Also please rename your issue to something more specific. Maybe "Improve typing module compatibility with types.Union" ?

File: Lib/test/test_typing.py [hunk 3]
```diff
@@ -1785,6 +1789,7 @@ def test_extended_generic_rules_repr(self):
     def test_generic_forward_ref(self):
         def foobar(x: List[List['CC']]): ...
         def foobar2(x: list[list[ForwardRef('CC')]]): ...
+        def foobar3(x: list[ForwardRef('CC')] | int): ...
```
Comment (Fidget-Spinner) [inline Lib/test/test_typing.py]: Please nest a union inside the ForwardRef too to test the `_eval_type` recursion is correct.

```suggestion
        def foobar3(x: list[ForwardRef('CC | int')] | int): ...
```
The assertion below will need updating as well.

**Review by Fidget-Spinner — COMMENTED** [review]
LGTM. Some minor questions and comments below.

File: Lib/typing.py (hunk 1 above)
Comment (uriyyo) [reply Lib/typing.py:1780]: I have added test to cover this case

**Comment by uriyyo** [comment]
@Fidget-Spinner Could you please review this PR?

**Review by Fidget-Spinner — APPROVED** [review]
Yurii, sorry for the delay. Was a little busy. This LGTM. Thanks for submitting this PR.

BTW, it looks like the C changes to types.Union in the other PR won't make to 3.10 (we missed the beta 4 deadline). So we don't have to worry about backporting anymore.

**Comment by uriyyo** [comment]
Thanks for review, should we mention those changes at `What's new` section?

**Comment by Fidget-Spinner** [comment]
> Thanks for review, should we mention those changes at `What's new` section?

Yes I was just thinking about that ;). I recommend we mention in whatsnew for 3.11 that union now supports nested type parameters.

**Comment by uriyyo** [comment]
Great, should it be done at separate PR?

**Comment by Fidget-Spinner** [comment]
> Great, should it be done at separate PR?

Please do! I don't think the other C changes to fix typing.Annotated are whatsnew worthy. So you can start on it.

**Comment by uriyyo** [comment]
@gvanrossum Could you please review this PR?

**Review by gvanrossum — APPROVED** [review]
Great work! I will merge.

**Comment by gvanrossum** [comment]
@Fidget-Spinner Do you think this is worth a backport to 3.10? IIUC @serhiy-storchaka is backporting his types.Union improvements.

**Comment by Fidget-Spinner** [comment]
I think it's worth a backport but we can't since the PR adding ``__parameters__`` https://github.com/python/cpython/pull/26980 was never backported.

**Comment by gvanrossum** [comment]
Never mind then. I guess that’s fine.

**Comment by Fidget-Spinner** [comment]
Thanks to Serhiy's Mass Cleanup, we can now backport this.

**Pull request merged**